README.md
# ember-cli-deploy-cloudformation

[![Build Status](https://github.com/kaliber5/ember-cli-deploy-cloudformation/workflows/CI/badge.svg)](https://github.com/kaliber5/ember-cli-deploy-cloudformation/actions)
[![Ember Observer Score](https://emberobserver.com/badges/ember-cli-deploy-cloudformation.svg)](https://emberobserver.com/addons/ember-cli-deploy-cloudformation)
[![npm version](https://badge.fury.io/js/ember-cli-deploy-cloudformation.svg)](https://badge.fury.io/js/ember-cli-deploy-cloudformation)

> An Ember CLI Deploy plugin to create/update an AWS CloudFormation stack before deploying to it

## Introduction

### What is CloudFormation?

[AWS CloudFormation][9] is Amazon's implementation of [infrastructure as code][6], which lets you create, update, provision
and delete stacks of AWS resources (e.g. EC2 instances, S3 buckets, CloudFront distributions etc.), described by a single 
configuration file.

For details on how CloudFormation works and how to write a template file, please visit the official
[CloudFormation Documentation][7]. Also check the [Examples section](#examples) for some example configurations for
common Ember deployment scenarios.

### What does this plugin do?

This is an Ember CLI addon that adds a plugin to the [Ember CLI Deploy][8] pipeline to create or update a CloudFormation
stack as part of the Ember CLI Deploy pipeline. It does not actually deploy any artefacts itself, this is where any of
the existing plugins that work with AWS resources come into play. Here are some that should work just fine:  

* [`ember-cli-deploy-s3`](https://github.com/ember-cli-deploy/ember-cli-deploy-s3)
* [`ember-cli-deploy-s3-index`](https://github.com/ember-cli-deploy/ember-cli-deploy-s3-index)
* [`ember-cli-deploy-s3-pack`](https://github.com/Gaurav0/ember-cli-deploy-s3-pack)
* [`ember-cli-deploy-cloudfront`](https://github.com/kpfefferle/ember-cli-deploy-cloudfront)
* [`ember-cli-deploy-elastic-beanstalk`](https://github.com/tomdale/ember-cli-deploy-elastic-beanstalk)
* [`ember-cli-deploy-fastboot-app-server-aws`](https://github.com/ember-cli-deploy/ember-cli-deploy-fastboot-app-server-aws)

This allows you to create the AWS resources and deploy your app to them in one single `ember deploy` command. 
For example you could easily deploy a feature branch to a new staging environment just for this feature and delete it 
afterwards, without any manual setup work.

### Why not use the AWS CLI directly?

The AWS CLI allows you to interact with CloudFormation, but suffers from a few caveats when integrating it into a
CI/CD pipeline:

* on the first run of the pipeline you need to create the stack, while you need to update it for all following
deployments.
* an update run without any changes to the stack is treated as an error by the CLI.
* there is no easy way to pass the outputs of the template (e.g. the name of a created S3 bucket) to any following
deployment steps.

## Installation

Run the following command in your terminal:

```bash
ember install ember-cli-deploy-cloudformation
```

## Examples

Ember is about shared solutions to common problems. And this applies here as well. Instead of figuring all the details
of a useful CloudFormation stack, this section will feature a few examples to get you going quickly:

* [S3 + CloudFront](examples/S3+CloudFront)
With a custom domain name (self hosted) and SSL

*This section is still a work in progess. If you have some templates you can share, please submit a PR so others can
benefit from your experience as well!*

## Configuration 

The configuration of this plugin takes place in three steps:

### 1. Create a CloudFormation template

Create a CloudFormation template, that defines all the AWS resources in your stack. Consult the 
[CloudFormation documentation][3] for more information.

Usually you will want to put this file under version control, so you can create it in your app's root folder as e.g.
`cfn.yaml`, `git add cfn-yaml`, and reference it in the plugin configuration as 
[`templateBody: 'file://cfn.yaml`](#templateBody).

### 2. Configure the plugin

Add the required configuration to your `config/deploy.js`.

See [Configuration Options](#configuration-options) below for all available configuration options. 
For detailed information on how configuration of plugins works, please refer to the [Plugin Documentation][1].

#### Example

Here is an example of a configuration. It uses environment variables to inject all parameters that are variable:

```js
{
  cloudformation: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    region: 'eu-central-1',
    templateBody: 'file://cfn.yaml',
    capabilities: ['CAPABILITY_IAM'],
    parameters: {
      DomainName: process.env.CFN_DOMAINNAME,
      Certificate: process.env.CFN_CERTIFICATE,
      HostedZoneId: process.env.CFN_HOSTEDZONEID
    }
  }
}
```

### 3. Pipe outputs to other plugins

A CloudFormation template let's you define so called "Outputs". These are commonly values, which were not available before 
creating the stack. So for example these could be some resource properties, like the name of an S3 bucket created by 
CloudFormation or the URL of a CloudFront distribution. 

Some of these you will have to pass to other Ember CLI Deploy plugins, so they can actually deploy your app to the
infrastructure that CloudFormation has created for you. A common example would be to pass the S3 bucket to 
`ember-cli-deploy-s3`.

This plugin will write all outputs of the processed template to the [Ember CLI Deploy context][4], at the path
`cloudformation.outputs` as simple hash with the name of the output as the key. As Ember CLI Deploy allow you to 
configure plugins not only with static values, but also with functions that receive the context, you can easily
pass the outputs to any other plugin. Here the afore-mentioned S3 plugin will receive the name of the bucket, that was
defined in the CloudFormation template output as `AssetsBucket`:

```js
{
  cloudformation: {
    ...
  },
  s3: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    bucket(context) {
      return context.cloudformation.outputs.AssetsBucket;
    },
    region: 'eu-central-1'
  }
}
```

### Configuration Options

#### accessKeyId

The AWS access key for the user that has the ability to work with CloudFormation. If this is left undefined,
the normal [AWS SDK credential resolution][5] will take place.

*Default:* `undefined`

#### secretAccessKey

The AWS secret for the user that has the ability to work with CloudFormation. This must be defined when `accessKeyId` is defined.

*Default:* `undefined`

#### profile

The AWS profile as definied in `~/.aws/credentials`. If this is left undefined,
the normal [AWS SDK credential resolution][5] will take place.

*Default:* `undefined`

#### region

The region to create the CloudFormation stack in.

*Default:* `undefined`

<hr>

Moreover all parameters supported by the AWS CloudFormation SDK will be passed when creating/updating the stack. See the
[AWS SDK docs][3] for a complete reference.

> Note: the plugin will let you write parameters in the more usual camelcase style, and map them to the pascal case 
style that AWS expects. So instead of writing `StackName`, you can use `stackName`. Also for the `tags` and `parameters`
properties, specify them as normal JavaScript hashes (e.g. `tags: { key: 'value' }`) instead of the original AWS syntax! 

Here is a list of the commonly used CloudFormation parameters:

#### stackName

The name of the CLoudFormation stack.

*Default:* `<appName>-<deployTarget>`, e.g. `myApp-production`

#### templateBody

The CloudFormation template body as a string (JSON or YAML). Instead of inlining your template, the `file://` protocol 
is also supported to reference a template file on disk, e.g. `file://path/to/cfn.yaml` (relative to your app's root folder)

Either `templateBody` or `templateURL` is **required**!

*Default:* `undefined`

#### templateURL

The location of a CloudFormation template (JSON or YAML). The URL must point to a template that is located in an Amazon S3 bucket.

Either `templateBody` or `templateURL` is **required**!

*Default:* `undefined`

#### capabilities

An array of required capabilites, e.g. `['CAPABILITY_IAM']` for working with IAM roles, user, policies etc.

*Default:* `undefined`

#### parameters

A hash of input parameters as expected by your CloudFormation template. 

*Note: specify them as normal JavaScript hashes (e.g. `parameters: { name: 'value' }`) instead of the unusual AWS syntax!*

```js
parameters: {
  Domain: 'example.com'
}
```

*Default:* `undefined`


[1]: http://ember-cli-deploy.com/plugins/ "Plugin Documentation"
[2]: https://aws.amazon.com/documentation/cloudformation/ "AWS CloudFormation Documentation"
[3]: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#createStack-property "AWS parameters"
[4]: http://ember-cli-deploy.com/docs/v1.0.x/the-deployment-context/ "Ember CLI Deploy context"
[5]: https://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html#Setting_AWS_Credentials "Setting AWS Credentials"
[6]: https://en.wikipedia.org/wiki/Infrastructure_as_Code
[7]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/Welcome.html
[8]: http://ember-cli-deploy.com/
[9]: https://aws.amazon.com/cloudformation/


CHANGELOG.md

## v1.0.0 (2022-06-02)

#### :boom: Breaking Change
* [#183](https://github.com/kaliber5/ember-cli-deploy-cloudformation/pull/183) Drop support for node 10 and 12 ([@simonihmig](https://github.com/simonihmig))

#### :house: Internal
* [#186](https://github.com/kaliber5/ember-cli-deploy-cloudformation/pull/186) Setup release-it ([@simonihmig](https://github.com/simonihmig))
* [#181](https://github.com/kaliber5/ember-cli-deploy-cloudformation/pull/181) Configure Renovate ([@renovate[bot]](https://github.com/apps/renovate))

#### Committers: 1
- Simon Ihmig ([@simonihmig](https://github.com/simonihmig))

# Changelog


tests/unit/cfn-test.js
/*eslint-env node*/
'use strict';

const fs = require('fs');
const chai = require('chai');
const chaiAsPromised = require("chai-as-promised");
const sinon = require('sinon');
const chaiSinon = require('sinon-chai');
const CfnClient = require('../../lib/cfn');
const AWS = require('aws-sdk');
const { expect } = chai;

chai.use(chaiAsPromised);
chai.use(chaiSinon);

const templatePath = 'tests/fixtures/cfn.yaml';

const options = {
  accessKeyId: 'abc',
  secretAccessKey: 'def',
  region: 'us-east-1',
  stackName: 'myStack',
  templateBody: `file://${templatePath}`,
  parameters: {
    key1: 'val1',
    key2: 'val2',
    key3: undefined
  },
  capabilities: ['CAPABILITY_IAM'],
  resourceTypes: ['AWS::*'],
  roleArn: 'ROLE',
  stackPolicyBody: 'body',
  notificationARNs: 'arn',
  tags: {
    key1: 'val1',
    key2: 'val2'
  },
  timeoutInMinutes: 10,
  disableRollback: true,
  rollbackConfiguration: {
    MonitoringTimeInMinutes: 10
  },
  dummy: undefined
};

const templateBody = fs.readFileSync(templatePath, { encoding: 'utf8' });

const expectedOptions = {
  StackName: 'myStack',
  TemplateBody: templateBody,
  Parameters: [
    {
      ParameterKey: 'key1',
      ParameterValue: 'val1'
    },
    {
      ParameterKey: 'key2',
      ParameterValue: 'val2'
    }
  ],
  Capabilities: ['CAPABILITY_IAM'],
  ResourceTypes: ['AWS::*'],
  RoleArn: 'ROLE',
  StackPolicyBody: 'body',
  NotificationARNs: 'arn',
  Tags: [
    {
      Key: 'key1',
      Value: 'val1'
    },
    {
      Key: 'key2',
      Value: 'val2'
    }
  ],
  TimeoutInMinutes: 10,
  DisableRollback: true,
  RollbackConfiguration: {
    MonitoringTimeInMinutes: 10
  }
};

const describeData = {
  Stacks: [{
    StackName: 'myStack',
    StackStatus: 'CREATE_COMPLETE',
    Outputs: [
      {
        OutputKey: 'AssetsBucket',
        OutputValue: 'abc-123456789'
      },
      {
        OutputKey: 'CloudFrontDistribution',
        OutputValue: 'EFG123456789'
      }
    ]
  }]
};

describe('Cloudformation client', function() {
  let client;
  let logger;

  beforeEach(function() {
    client = new CfnClient(options);
    sinon.stub(client.awsClient, 'createStack').returns({
      promise: sinon.fake.resolves()
    });
    sinon.stub(client.awsClient, 'updateStack').returns({
      promise: sinon.fake.resolves()
    });
    sinon.stub(client.awsClient, 'waitFor').returns({
      promise: sinon.fake.resolves()
    });
    sinon.stub(client.awsClient, 'describeStacks').returns({
      promise: sinon.fake.resolves(describeData)
    });
    sinon.stub(client.awsClient, 'validateTemplate').returns({
      promise: sinon.fake.resolves()
    });

    // minimal logger interface
    logger = {
      log: sinon.fake(),
      error: sinon.fake(),
      debug: sinon.fake()
    };

    client.logger = logger;
  });

  afterEach(function() {
    sinon.restore();
  });

  it('passes constructor args to AWS.CloudFormation', function() {
    let constructor = sinon.fake();
    sinon.replace(AWS, 'CloudFormation', constructor);

    new CfnClient({
      foo: 'bar',
      accessKeyId: 'abc',
      secretAccessKey: 'def',
      region: 'us-east-1'
    });

    expect(constructor).to.always.have.been.calledWithNew;
    expect(constructor).to.have.been.calledWith({
      apiVersion: '2010-05-15',
      accessKeyId: 'abc',
      secretAccessKey: 'def',
      region: 'us-east-1'
    });
  });

  function checkCreateStack(callFn) {
    it('it calls createStack with adjusted options', async function() {
      await expect(callFn()).to.be.fulfilled;
      expect(client.awsClient.createStack).to.have.been.calledWith(expectedOptions);
    });

    it('it waits for stackCreateComplete', async function() {
      await expect(callFn()).to.be.fulfilled;
      expect(client.awsClient.waitFor).to.have.been.calledWith('stackCreateComplete', { StackName: 'myStack' });
      expect(logger.debug).to.have.been.calledWith(`Creating new CloudFormation stack 'myStack'...`);
      expect(logger.log).to.have.been.calledWith(`New CloudFormation stack 'myStack' has been created!`);
      expect(logger.debug).to.have.been.calledBefore(logger.log);
    });

    it('rejects when createStack fails', function() {
      client.awsClient.createStack.returns({
        promise: sinon.fake.rejects()
      });

      return expect(callFn()).to.be.rejected;
    });
  }

  function checkUpdateStack(callFn) {
    it('it calls updateStack with adjusted options', async function() {
      await expect(callFn()).to.be.fulfilled;
      expect(client.awsClient.updateStack).to.have.been.calledWith(expectedOptions);
    });

    it('it waits for stackUpdateComplete', async function() {
      await expect(callFn()).to.be.fulfilled;
      expect(client.awsClient.waitFor).to.have.been.calledWith('stackUpdateComplete', { StackName: 'myStack' });
      expect(logger.debug).to.have.been.calledWith(`Updating CloudFormation stack 'myStack'...`);
      expect(logger.log).to.have.been.calledWith(`CloudFormation stack 'myStack' has been updated!`);
      expect(logger.debug).to.have.been.calledBefore(logger.log);
    });

    it('rejects when updateStack fails', function() {
      client.awsClient.updateStack.returns({
        promise: sinon.fake.rejects()
      });

      return expect(callFn()).to.be.rejected;
    });

    it('ignores unchanged stack', async function() {
      client.awsClient.updateStack.returns({
        promise: sinon.fake.rejects('No updates are to be performed')
      });

      await expect(callFn()).to.be.fulfilled;
      expect(client.awsClient.waitFor).to.not.have.been.called;
      expect(logger.debug).to.have.been.calledWith(`No updates are to be performed to CloudFormation stack 'myStack'`);
    });
  }

  describe('createStack', function() {
    checkCreateStack(() => client.createStack());
  });

  describe('updateStack', function() {
    checkUpdateStack(() => client.updateStack());
  });

  describe('validateTemplate', function() {
    it('resolves for valid template', async function() {
      await expect(client.validateTemplate()).to.be.fulfilled;
      expect(client.awsClient.validateTemplate).to.have.been.calledWith({
        TemplateBody: templateBody
      })
    });

    it('rejects for invalid template', async function() {
      client.awsClient.validateTemplate.returns({
        promise: sinon.fake.rejects('template error')
      });
      await expect(client.validateTemplate()).to.be.rejected;
      expect(client.awsClient.validateTemplate).to.have.been.calledWith({
        TemplateBody: templateBody
      })
    });
  });

  describe('stackExists', function() {
    it('returns true when stack exists', function() {
      return expect(client.stackExists()).to.eventually.be.true;
    });

    it('returns false when stack does not exists', function() {
      client.awsClient.describeStacks.returns({
        promise: sinon.fake.rejects('Stack with id myStack does not exist')
      });
      return expect(client.stackExists()).to.eventually.be.false;
    });
  });

  describe('createOrUpdateStack', function() {
    describe('stack does not exist', function() {
      beforeEach(function() {
        sinon.stub(client, 'stackExists').resolves(false);
      });

      checkCreateStack(() => client.createOrUpdateStack());
    });

    describe('stack exists', function() {
      beforeEach(function() {
        sinon.stub(client, 'stackExists').resolves(true);
      });

      checkUpdateStack(() => client.createOrUpdateStack());
    });
  });

  describe('fetchOutputs', function() {
    it('returns adjusted output hash', function() {
      return expect(client.fetchOutputs()).to.eventually.deep.equal({
        AssetsBucket: 'abc-123456789',
        CloudFrontDistribution: 'EFG123456789'
      });
    });

    it('returns empty hash when no outputs are found', function() {
      let emptyDescribeData = JSON.parse(JSON.stringify(describeData));
      delete emptyDescribeData.Stacks[0].Outputs;
      client.awsClient.describeStacks.returns({
        promise: sinon.fake.resolves(emptyDescribeData)
      });

      return expect(client.fetchOutputs()).to.eventually.deep.equal({});
    });
  });

});


tests/helpers/assert.js
/*eslint-env node*/
const chai = require('chai');
const chaiAsPromised = require("chai-as-promised");
chai.use(chaiAsPromised);

module.exports = chai.assert;


tests/unit/index-test.js
/*eslint-env node*/
'use strict';

const CfnClient = require('../../lib/cfn');
const chai = require('chai');
const chaiAsPromised = require("chai-as-promised");
const chaiSinon = require('sinon-chai');
const subject = require('../../index');
const sinon = require('sinon');
const { expect } = chai;

chai.use(chaiAsPromised);
chai.use(chaiSinon);

const expectedOutputs = {
  AssetsBucket: 'abc-123456789',
  CloudFrontDistribution: 'EFG123456789'
};

const pluginOptions = {
  templateBody: 'dummy',
  region: 'eu-central-1'
};

describe('Cloudformation plugin', function() {
  let mockUi;
  let instance;
  let cfnClient;
  let context;

  beforeEach(function() {
    mockUi = {
      verbose: true,
      messages: [],
      write() {
      },
      writeLine(message) {
        this.messages.push(message);
      }
    };

    context = {
      ui: mockUi,
      config: {
        cloudformation: Object.assign({}, pluginOptions)
      },
      project: {
        name() {
          return 'myapp'
        }
      },
      deployTarget: 'production'
    };

    instance = subject.createDeployPlugin({
      name: 'cloudformation'
    });

    cfnClient = new CfnClient(context.config.cloudformation);
    context.config.cloudformation.cfnClient = cfnClient;
    sinon.stub(cfnClient, 'validateTemplate').resolves();
    sinon.stub(cfnClient, 'createOrUpdateStack').resolves();
    sinon.stub(cfnClient, 'fetchOutputs').resolves(expectedOutputs);
  });

  afterEach(function() {
    sinon.restore();
  });

  it('has a name', function() {
    let instance = subject.createDeployPlugin({
      name: 'cloudformation'
    });

    expect(instance.name).to.equal('cloudformation');
  });

  it('implements the correct hooks', function() {
    let plugin = subject.createDeployPlugin({
      name: 'cloudformation'
    });

    expect(plugin.configure).to.be.a('function');
    expect(plugin.prepare).to.be.a('function');
  });

  describe('configure', function() {
    describe('required config', function() {
      it('errors on missing config props', function() {
        delete context.config.cloudformation.templateBody;
        instance.beforeHook(context);

        let msg = `Missing required config: either 'templateBody' or 'templateURL' is required`;

        expect(() => instance.configure(context)).to.throw(msg);
        expect(mockUi.messages.pop()).to.include(msg);
      });
    });

    describe('default config', function() {
      it('provides default stackName', function() {
        instance.beforeHook(context);
        instance.configure(context);

        expect(instance.readConfig('stackName')).to.equal('myapp-production');
      });
    });
  });

  describe('setup', function() {
    it('resolves', function() {
      instance.beforeHook(context);
      return expect(instance.setup(context)).to.be.fulfilled;
    });

    it('errors on invalid template', async function() {
      let msg = 'invalid template';
      cfnClient.validateTemplate.rejects(msg);

      instance.beforeHook(context);

      await expect(instance.setup(context)).to.be.rejected;
      expect(mockUi.messages.pop()).to.include(msg);
    });

  });

  describe('prepare', function() {
    it('creates or updates stack', async function() {
      instance.beforeHook(context);
      instance.setup(context);
      await expect(instance.prepare(context)).to.be.fulfilled;
      expect(cfnClient.createOrUpdateStack).to.have.been.calledOnce;
    });

    it('adds outputs to context', async function() {
      instance.beforeHook(context);
      instance.setup(context);
      await expect(instance.prepare(context)).to.be.fulfilled;
      expect(context.cloudformation).to.exist;
      expect(context.cloudformation).to.have.property('outputs', expectedOutputs);
    });
  });

});


index.js
/*eslint-env node*/
'use strict';

//const RSVP = require('rsvp');
const DeployPluginBase = require('ember-cli-deploy-plugin');
const CfnClient = require('./lib/cfn');
const Logger = require('./lib/logger');

module.exports = {
  name: 'ember-cli-deploy-cloudformation',

  createDeployPlugin: function(options) {
    let DeployPlugin = DeployPluginBase.extend({
      name: options.name,

      defaultConfig: {},
      requiredConfig: [],

      configure(context) {
        this.defaultConfig.stackName = `${context.project.name()}-${context.deployTarget}`;
        this._super.configure.apply(this, arguments);

        let templateBody = this.readConfig('templateBody');
        let templateUrl = this.readConfig('templateURL');

        if (!templateBody && !templateUrl) {
          let message = `Missing required config: either 'templateBody' or 'templateURL' is required`;
          this.log(message, { color: 'red' });
          throw new Error(message);
        }
      },

      setup() {
        let options = Object.keys(this.pluginConfig)
          .map((key) => ({ [key]: this.readConfig(key)}))
          .reduce((result, item) => Object.assign(result, item), {});

        this.cfnClient = this.readConfig('cfnClient') || new CfnClient(options);
        this.cfnClient.logger = new Logger(this.log.bind(this));

        return this.cfnClient.validateTemplate()
          .catch(this._errorMessage.bind(this));
      },

      prepare(context) {
        return this.cfnClient.createOrUpdateStack()
          .then(() => this.cfnClient.fetchOutputs())
          .then(outputs => {
            context.cloudformation = {
              outputs
            };
          })
          .catch(this._errorMessage.bind(this));
      },

      _errorMessage(error) {
        this.log(error, { color: 'red' });
        if (error && error.stack) {
          this.log(error.stack, { color: 'red' });
        }
        return Promise.reject(error);
      }
    });

    return new DeployPlugin();
  }
};


examples/S3+CloudFront/deploy.js
/* eslint-env node */
'use strict';

module.exports = function(deployTarget) {
  let ENV = {
    build: {
      environment: 'production'
    },
    pipeline: {
      // This setting runs the ember-cli-deploy activation hooks on every deploy
      // which is necessary in order to run ember-cli-deploy-cloudfront.
      // To disable CloudFront invalidation, remove this setting or change it to `false`.
      // To disable ember-cli-deploy-cloudfront for only a particular environment, add
      // `ENV.pipeline.activateOnDeploy = false` to an environment conditional below.
      activateOnDeploy: true
    },
    cloudformation: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID,             // SET THE ENV VAR
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,     // SET THE ENV VAR
      region: 'us-east-1',
      stackName: `${require('../package.json').name}-${deployTarget}`,
      templateBody: 'file://cfn.yaml',
      capabilities: ['CAPABILITY_IAM'],
      parameters: {
        DomainName: process.env.CFN_DOMAINNAME,               // SET THE ENV VAR OR REPLACE!
        CFCertificate: process.env.CFN_CFCERTIFICATE          // SET THE ENV VAR OR REPLACE!
      }
    },
    s3: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
      bucket(context) {
        return context.cloudformation.outputs.AssetsBucket;
      },
      region: 'us-east-1'
    },
    's3-index': {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
      bucket(context) {
        return context.cloudformation.outputs.AssetsBucket;
      },
      region: 'us-east-1',
      filePattern: 'index.html'
    },
    cloudfront: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
      distribution(context) {
        return context.cloudformation.outputs.CloudFrontDistribution;
      }
    }
  };

  // Note: if you need to build some configuration asynchronously, you can return
  // a promise that resolves with the ENV object instead of returning the
  // ENV object synchronously.
  return ENV;
};


lib/cfn.js
'use strict';

const AWS = require('aws-sdk');
const fs = require('fs');

function hash2ArrayHash(hash, keyProperty = 'Key', valueProperty = 'Value') {
  return Object.keys(hash)
    .map(key => {
      return hash[key] !== undefined ? {
        [keyProperty]: key,
        [valueProperty]: hash[key]
      } : undefined;
    })
    .filter(Boolean);
}

function ucFirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

class CfnClient {

  constructor(options) {

    let awsOptions = {
      apiVersion: '2010-05-15',
      region: options.region,
      accessKeyId: options.accessKeyId,
      secretAccessKey: options.secretAccessKey
    };

    if (options.profile) {
      awsOptions.credentials = new AWS.SharedIniFileCredentials({ profile: options.profile });
    }

    let cfnOptions = Object.assign({}, options);
    delete cfnOptions.region;
    delete cfnOptions.accessKeyId;
    delete cfnOptions.secretAccessKey;
    delete cfnOptions.profile;

    this.setOptions(cfnOptions);
    this.awsClient = new AWS.CloudFormation(awsOptions);
  }

  validateTemplate() {
    let options = {};

    if (this.awsOptions.TemplateBody) {
      options.TemplateBody = this.awsOptions.TemplateBody
    } else {
      options.TemplateURL = this.awsOptions.templateURL;
    }

    return this.awsClient.validateTemplate(options).promise();
  }

  async stackExists() {
    try {
      await this.awsClient
        .describeStacks({ StackName: this.options.stackName })
        .promise();
      return true;
    } catch(err) {
      if (String(err).endsWith('does not exist')) {
        return false;
      }
      throw err;
    }
  }

  async createStack() {
    this.log(`Creating new CloudFormation stack '${this.options.stackName}'...`, 'debug');

    await this.awsClient
      .createStack(this.awsOptions)
      .promise();
    await this.awsClient.waitFor('stackCreateComplete', { StackName: this.options.stackName }).promise();

    this.log(`New CloudFormation stack '${this.options.stackName}' has been created!`)
  }

  async updateStack() {
    this.log(`Updating CloudFormation stack '${this.options.stackName}'...`, 'debug');
    try {
      await this.awsClient
        .updateStack(this.awsOptions)
        .promise();
      await this.awsClient.waitFor('stackUpdateComplete', { StackName: this.options.stackName }).promise();
      this.log(`CloudFormation stack '${this.options.stackName}' has been updated!`);
    } catch(err) {
      if (String(err).includes('No updates are to be performed')) {
        this.log(`No updates are to be performed to CloudFormation stack '${this.options.stackName}'`, 'debug');
        return;
      }
      throw err;
    }
  }

  async createOrUpdateStack() {
    let result = await this.stackExists()
    if (result === true) {
      return this.updateStack();
    }
    return this.createStack();
  }

  async fetchOutputs() {
    let result = await this.awsClient
      .describeStacks({ StackName: this.options.stackName })
      .promise();

    if (!result.Stacks || !result.Stacks[0]) {
      throw new Error('No stack data found from `describeStacks` call');
    }

    let data = result.Stacks[0];

    if (!data.Outputs) {
      return {};
    }

    return data.Outputs
      .map(item => ({ [item.OutputKey]: item.OutputValue }))
      .reduce((result, item) => Object.assign(result, item), {})
  }

  setOptions(options) {
    this.options = options;

    this.awsOptions = Object.keys(options)
      .map(origKey => {
        let value = options[origKey];
        let key = ucFirst(origKey);

        if (value === undefined) {
          return;
        }

        switch (origKey) {
          case 'parameters':
            value = hash2ArrayHash(value, 'ParameterKey', 'ParameterValue');
            break;
          case 'tags':
            value = hash2ArrayHash(value);
            break;
          case 'templateBody':
            if (value.startsWith('file://')) {
              value = fs.readFileSync(value.slice(7), { encoding: 'utf8' })
            }
        }

        return { [key]: value };
      })
      .reduce((result, item) => Object.assign(result, item), {});
  }

  log(message, type = 'log') {
    if (this.logger) {
      if (typeof this.logger[type] !== 'function') {
        throw new Error(`Logger does not implement ${type} type`);
      }
      this.logger[type](message);
    }
  }

}

module.exports = CfnClient;
